File: Sources/Chromabase/Adjustments/LocalDodgeBurn.swift
import Foundation

public struct LocalDodgeBurnPoint: Codable, Sendable, Equatable {
    public var x: Double
    public var y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

public struct LocalDodgeBurnStroke: Codable, Sendable, Equatable {
    public var points: [LocalDodgeBurnPoint]
    public var thickness: Double
    public var feather: Double

    public init(points: [LocalDodgeBurnPoint], thickness: Double = 0.04, feather: Double = 0.02) {
        self.points = points
        self.thickness = thickness
        self.feather = feather
    }

    enum CodingKeys: String, CodingKey { case points, thickness, feather }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        points = try c.decodeIfPresent([LocalDodgeBurnPoint].self, forKey: .points) ?? []
        thickness = try c.decodeIfPresent(Double.self, forKey: .thickness) ?? 0.04
        feather = try c.decodeIfPresent(Double.self, forKey: .feather) ?? 0.02
    }
}

public enum LocalDodgeBurnMode: String, Codable, Sendable, Equatable, Hashable, CaseIterable {
    case dodge
    case burn
}

public struct LocalDodgeBurnMask: Codable, Sendable, Equatable {
    public enum Kind: String, Codable, Sendable, CaseIterable {
        case brush
        case radial
        case linear
        case polygon
    }

    public var kind: Kind
    public var strokes: [LocalDodgeBurnStroke]
    public var center: LocalDodgeBurnPoint
    public var radius: Double
    public var feather: Double
    public var start: LocalDodgeBurnPoint
    public var end: LocalDodgeBurnPoint
    public var points: [LocalDodgeBurnPoint]

    public static func brush(strokes: [LocalDodgeBurnStroke]) -> LocalDodgeBurnMask {
        LocalDodgeBurnMask(kind: .brush, strokes: strokes)
    }

    public static func radial(center: LocalDodgeBurnPoint, radius: Double, feather: Double) -> LocalDodgeBurnMask {
        LocalDodgeBurnMask(kind: .radial, center: center, radius: radius, feather: feather)
    }

    public static func linear(
        start: LocalDodgeBurnPoint,
        end: LocalDodgeBurnPoint,
        feather: Double
    ) -> LocalDodgeBurnMask {
        LocalDodgeBurnMask(kind: .linear, feather: feather, start: start, end: end)
    }

    public static func polygon(points: [LocalDodgeBurnPoint], feather: Double) -> LocalDodgeBurnMask {
        LocalDodgeBurnMask(kind: .polygon, feather: feather, points: points)
    }

    public init(
        kind: Kind,
        strokes: [LocalDodgeBurnStroke] = [],
        center: LocalDodgeBurnPoint = LocalDodgeBurnPoint(x: 0.5, y: 0.5),
        radius: Double = 0.25,
        feather: Double = 0.25,
        start: LocalDodgeBurnPoint = LocalDodgeBurnPoint(x: 0.5, y: 0.0),
        end: LocalDodgeBurnPoint = LocalDodgeBurnPoint(x: 0.5, y: 1.0),
        points: [LocalDodgeBurnPoint] = []
    ) {
        self.kind = kind
        self.strokes = strokes
        self.center = center
        self.radius = radius
        self.feather = feather
        self.start = start
        self.end = end
        self.points = points
    }

    enum CodingKeys: String, CodingKey {
        case kind, strokes, center, radius, feather, start, end, points
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        kind = try c.decode(Kind.self, forKey: .kind)
        strokes = try c.decodeIfPresent([LocalDodgeBurnStroke].self, forKey: .strokes) ?? []
        center = try c.decodeIfPresent(LocalDodgeBurnPoint.self, forKey: .center) ?? LocalDodgeBurnPoint(x: 0.5, y: 0.5)
        radius = try c.decodeIfPresent(Double.self, forKey: .radius) ?? 0.25
        feather = try c.decodeIfPresent(Double.self, forKey: .feather) ?? 0.25
        start = try c.decodeIfPresent(LocalDodgeBurnPoint.self, forKey: .start) ?? LocalDodgeBurnPoint(x: 0.5, y: 0.0)
        end = try c.decodeIfPresent(LocalDodgeBurnPoint.self, forKey: .end) ?? LocalDodgeBurnPoint(x: 0.5, y: 1.0)
        points = try c.decodeIfPresent([LocalDodgeBurnPoint].self, forKey: .points) ?? []
    }
}

public struct LocalDodgeBurnAdjustment: Codable, Sendable, Equatable, Identifiable {
    public var id: UUID
    public var mode: LocalDodgeBurnMode
    public var amount: Double
    public var isEnabled: Bool
    public var mask: LocalDodgeBurnMask

    public init(
        id: UUID = UUID(),
        mode: LocalDodgeBurnMode,
        amount: Double,
        isEnabled: Bool = true,
        mask: LocalDodgeBurnMask
    ) {
        self.id = id
        self.mode = mode
        self.amount = amount
        self.isEnabled = isEnabled
        self.mask = mask
    }

    enum CodingKeys: String, CodingKey { case id, mode, amount, isEnabled, mask }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        mode = try c.decodeIfPresent(LocalDodgeBurnMode.self, forKey: .mode) ?? .dodge
        amount = try c.decodeIfPresent(Double.self, forKey: .amount) ?? 0
        isEnabled = try c.decodeIfPresent(Bool.self, forKey: .isEnabled) ?? true
        mask = try c.decode(LocalDodgeBurnMask.self, forKey: .mask)
    }
}

File: negaflow-windows/src/Native/imaging/local_dodge_burn.cpp
#include "negaflow/imaging/local_dodge_burn.h"

#include "local_dodge_burn_masks.h"
#include "local_dodge_burn_types.h"

#include "negaflow/core/parallel_rows.h"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <new>
#include <utility>
#include <vector>

namespace negaflow::imaging {
namespace {

using namespace negaflow::imaging::local_dodge_burn_detail;

void discard_pixels(WorkingImage& image) noexcept {
    std::vector<negaflow::core::Rgba32F>{}.swap(image.pixels);
}

[[nodiscard]] negaflow::core::ConstImageView const_view(
    const WorkingImage& image) noexcept {
    return {
        image.pixels.data(),
        image.pixels.size(),
        image.width,
        image.height,
        image.stride_pixels,
    };
}


void apply_adjustment(
    WorkingImage& image,
    const LocalDodgeBurnAdjustment& adjustment,
    const std::vector<float>& mask) noexcept {
    const float amount = clamp_unit(adjustment.amount);
    const float stops = adjustment.mode == LocalDodgeBurnMode::dodge
        ? amount
        : -amount;
    const float exposure = std::exp2(stops);
    negaflow::core::for_each_row_block(
        image.height,
        static_cast<std::uint64_t>(image.width) *
            static_cast<std::uint64_t>(image.height),
        [&](const std::uint32_t first_row, const std::uint32_t row_count) noexcept {
      for (std::uint32_t y = first_row; y < first_row + row_count; ++y) {
        auto* const row = image.pixels.data() +
            static_cast<std::size_t>(y) * image.stride_pixels;
        for (std::uint32_t x = 0U; x < image.width; ++x) {
            const float weight = mask[index_of(x, y, image.width)];
            if (weight <= 0.0F) {
                continue;
            }
            const float scale = 1.0F + (exposure - 1.0F) * weight;
            row[x].red *= scale;
            row[x].green *= scale;
            row[x].blue *= scale;
        }
      }
        });
}

[[nodiscard]] bool finite_point(const LocalDodgeBurnPoint point) noexcept {
    return std::isfinite(point.x) && std::isfinite(point.y);
}

}  // namespace

bool valid_local_dodge_burn_parameters(
    const LocalDodgeBurnParameters& parameters) noexcept {
    if (parameters.adjustments.size() >
        local_dodge_burn_maximum_adjustments) {
        return false;
    }
    std::size_t total_points = 0U;
    for (const LocalDodgeBurnAdjustment& adjustment : parameters.adjustments) {
        if (!std::isfinite(adjustment.amount) ||
            static_cast<std::uint8_t>(adjustment.mode) >
                static_cast<std::uint8_t>(LocalDodgeBurnMode::burn) ||
            static_cast<std::uint8_t>(adjustment.mask.kind) >
                static_cast<std::uint8_t>(LocalDodgeBurnMaskKind::polygon) ||
            !finite_point(adjustment.mask.center) ||
            !finite_point(adjustment.mask.start) ||
            !finite_point(adjustment.mask.end) ||
            !std::isfinite(adjustment.mask.radius) ||
            !std::isfinite(adjustment.mask.feather) ||
            adjustment.mask.strokes.size() >
                local_dodge_burn_maximum_strokes_per_mask) {
            return false;
        }
        if (adjustment.mask.points.size() >
            local_dodge_burn_maximum_points - total_points) {
            return false;
        }
        total_points += adjustment.mask.points.size();
        for (const LocalDodgeBurnPoint point : adjustment.mask.points) {
            if (!finite_point(point)) {
                return false;
            }
        }
        for (const LocalDodgeBurnStroke& stroke : adjustment.mask.strokes) {
            if (!std::isfinite(stroke.thickness) ||
                !std::isfinite(stroke.feather) ||
                stroke.points.size() >
                    local_dodge_burn_maximum_points - total_points) {
                return false;
            }
            total_points += stroke.points.size();
            for (const LocalDodgeBurnPoint point : stroke.points) {
                if (!finite_point(point)) {
                    return false;
                }
            }
        }
    }
    return true;
}

LocalDodgeBurnResult apply_local_dodge_burn(
    WorkingImage image,
    const LocalDodgeBurnParameters& parameters) noexcept {
    LocalDodgeBurnResult result{};
    result.image = std::move(image);
    if (!valid_local_dodge_burn_parameters(parameters)) {
        discard_pixels(result.image);
        return result;
    }
    result.info.kernel_status =
        negaflow::core::validate_finite_pixels(const_view(result.image));
    if (result.info.kernel_status != negaflow::core::KernelStatus::ok) {
        result.status = LocalDodgeBurnStatus::kernel_failed;
        discard_pixels(result.image);
        return result;
    }
    if (parameters.adjustments.empty()) {
        result.status = LocalDodgeBurnStatus::ok;
        return result;
    }

    try {
        for (const LocalDodgeBurnAdjustment& adjustment :
             parameters.adjustments) {
            if (!adjustment.enabled ||
                clamp_unit(adjustment.amount) <=
                    adjustment_identity_threshold) {
                continue;
            }
            MaskResult mask = make_mask(adjustment.mask, result.image);
            result.info.mask_scratch_peak_bytes = std::max(
                result.info.mask_scratch_peak_bytes,
                mask.scratch_peak_bytes);
            if (mask.weights.empty() || !mask_has_weight(mask.weights)) {
                continue;
            }
            apply_adjustment(result.image, adjustment, mask.weights);
            ++result.info.adjustments_applied;
        }
        result.info.applied = result.info.adjustments_applied != 0U;
        result.status = LocalDodgeBurnStatus::ok;
        return result;
    } catch (const std::bad_alloc&) {
        result.status = LocalDodgeBurnStatus::allocation_failed;
        discard_pixels(result.image);
        return result;
    } catch (...) {
        result.status = LocalDodgeBurnStatus::allocation_failed;
        discard_pixels(result.image);
        return result;
    }
}

const char* local_dodge_burn_status_name(
    const LocalDodgeBurnStatus status) noexcept {
    switch (status) {
        case LocalDodgeBurnStatus::ok:
            return "ok";
        case LocalDodgeBurnStatus::invalid_parameter:
            return "invalid_parameter";
        case LocalDodgeBurnStatus::kernel_failed:
            return "kernel_failed";
        case LocalDodgeBurnStatus::allocation_failed:
            return "allocation_failed";
    }
    return "unknown";
}


}  // namespace negaflow::imaging
